### ISSUE TITLE
Incorrect autodiff Jacobian

### ISSUE THREAD

**Issue by perlinm** [jmlarson1/QFI-Opt#36]
I added a manual (finite difference) calculation of the Jacobian for JAX-less simulations in PR #34, and noticed that the manual Jacobian and autodiff/JAX one disagreed.  I **_think_** the autodiff/JAX one is incorrect.

Working on the `compare-jacobians` branch, compare (say) the leftmost column of the imaginary part of `d(final_state/d(params[1])`...
- as computed with JAX:
```
$ export DISABLE_JAX=; python3 qfi_opt/spin_models.py --num_qubits 2 --jacobian --params 0.5 0 0 0.5

final state:
[[ 0.25+0.j    0.  +0.25j  0.  +0.25j -0.25+0.j  ]
 [ 0.  -0.25j  0.25+0.j    0.25+0.j    0.  +0.25j]
 [ 0.  -0.25j  0.25+0.j    0.25+0.j    0.  +0.25j]
 [-0.25+0.j    0.  -0.25j  0.  -0.25j  0.25+0.j  ]]

d(final_state/d(params[1]):
real:
[[ 6.18e-08  7.85e-01  7.85e-01  1.43e-15]
 [ 7.85e-01 -1.37e-15 -1.37e-15  7.85e-01]
 [ 7.85e-01 -1.37e-15 -1.37e-15  7.85e-01]
 [ 1.43e-15  7.85e-01  7.85e-01 -6.18e-08]]
imag:
[[-1.57e+00  3.09e-08  3.09e-08 -3.52e-08]
 [ 3.09e-08  3.37e-08  3.37e-08  3.09e-08]
 [ 3.09e-08  3.37e-08  3.37e-08  3.09e-08]
 [-3.52e-08  3.09e-08  3.09e-08  1.57e+00]]
```
- versus that computed with finite differences:
```
$ export DISABLE_JAX=1; python3 qfi_opt/spin_models.py --num_qubits 2 --jacobian --params 0.5 0 0 0.5

final state:
[[ 0.25+0.j    0.  +0.25j  0.  +0.25j -0.25+0.j  ]
 [ 0.  -0.25j  0.25+0.j    0.25+0.j    0.  +0.25j]
 [ 0.  -0.25j  0.25+0.j    0.25+0.j    0.  +0.25j]
 [-0.25+0.j    0.  -0.25j  0.  -0.25j  0.25+0.j  ]]

d(final_state/d(params[1]):
real:
[[ 2.80e-04  7.85e-01  7.85e-01  2.47e-04]
 [ 7.85e-01 -2.47e-04 -2.47e-04  7.85e-01]
 [ 7.85e-01 -2.47e-04 -2.47e-04  7.85e-01]
 [ 2.47e-04  7.85e-01  7.85e-01  2.14e-04]]
imag:
[[ 0.00e+00 -1.07e-04 -1.07e-04  1.57e+00]
 [ 1.07e-04  0.00e+00  0.00e+00 -1.40e-04]
 [ 1.07e-04  0.00e+00  0.00e+00 -1.40e-04]
 [-1.57e+00  1.40e-04  1.40e-04  0.00e+00]]
```
The rows seem to be in reversed order.  Moreover, it looks to me like the non-JAX version is correct, since if I take a small step of `0.01` in the first parameter, the parameter on the bottom left of `final_state` picks up a small negative imaginary value:
```
$ python3 qfi_opt/spin_models.py --num_qubits 2 --jacobian --params 0.5 0.01 0 0.5

final state:
[[ 0.25+0.j    0.01+0.25j  0.01+0.25j -0.25+0.02j]
 [ 0.01-0.25j  0.25+0.j    0.25+0.j    0.01+0.25j]
 [ 0.01-0.25j  0.25+0.j    0.25+0.j    0.01+0.25j]
 [-0.25-0.02j  0.01-0.25j  0.01-0.25j  0.25+0.j  ]]
```

@sriharikrishna

**Comment by sriharikrishna** [comment]
I will check this will JAX's utility functions

**Comment by perlinm** [comment]
Simpler example, on the `compare-jacobians` branch:
- With JAX:
```bash
$ export DISABLE_JAX=; python3 qfi_opt/spin_models.py --num_qubits 1 --jacobian --params 0 0.5 0 0
final state:
[[0.5+0.j  0. -0.5j]
 [0. +0.5j 0.5+0.j ]]

d(final_state/d(params[1]):
real:
[[ 1.57  0.  ]
 [ 0.   -1.57]]
imag:
[[ 0.   -1.57]
 [-1.57  0.  ]]
```
- With finite differences:
```bash
$ export DISABLE_JAX=1; python3 qfi_opt/spin_models.py --num_qubits 1 --jacobian --params 0 0.5 0 0
final state:
[[0.5+0.j  0. -0.5j]
 [0. +0.5j 0.5+0.j ]]

d(final_state/d(params[1]):
real:
[[ 1.57  0.  ]
 [ 0.   -1.57]]
imag:
[[ 0.  0.]
 [-0.  0.]]
```

The only relevant part of `qfi_opt/spin_models.py` is the following:
![image](https://github.com/jmlarson1/QFI-Opt/assets/567092/6b4fc351-588c-4901-95fc-bfd81912b3f0)
where essentially `KET_0 = [1, 0]` and `KET_1 = [0, 1]`.

According to Mathematica, the derivative obtained by finite differences is correct:
![image](https://github.com/jmlarson1/QFI-Opt/assets/567092/e795bcb9-0160-4dd9-ac70-7500fa5bd7a7)

**Comment by sriharikrishna** [comment]
For commit `432d440e1fbeabfa295056d87a59041560410aa3`  setting `USE_JAX = False`, I tried both order of parameters for the canonical test and run into this issue:

```
(qfi_310) snarayan@CSI0362938 QFI-Opt % python spin_models.py --num_qubits 2 --params 0.0 0.5 0.0 0.0 
Traceback (most recent call last):
  File "/Users/snarayan/soft/QFI-Opt/spin_models.py", line 326, in <module>
    final_state = simulate_OAT(args.params, args.num_qubits, args.dissipation)
  File "/Users/snarayan/soft/QFI-Opt/spin_models.py", line 112, in simulate_OAT
    return simulate_sensing_protocol(params, hamiltonian, dissipation_rates, dissipation_format, axial_symmetry=True)
  File "/Users/snarayan/soft/QFI-Opt/spin_models.py", line 98, in simulate_sensing_protocol
    state_3 = evolve_state(state_2, time_3, final_hamiltonian)
  File "/Users/snarayan/soft/QFI-Opt/spin_models.py", line 219, in evolve_state
    final_vec = solution.y[:, -1]
TypeError: list indices must be integers or slices, not tuple
(qfi_310) snarayan@CSI0362938 QFI-Opt % 
```

and

```
(qfi_310) snarayan@CSI0362938 QFI-Opt % python spin_models.py --num_qubits 2 --params 0.5 0.0 0.0 0.0 
Traceback (most recent call last):
  File "/Users/snarayan/soft/QFI-Opt/spin_models.py", line 326, in <module>
    final_state = simulate_OAT(args.params, args.num_qubits, args.dissipation)
  File "/Users/snarayan/soft/QFI-Opt/spin_models.py", line 112, in simulate_OAT
    return simulate_sensing_protocol(params, hamiltonian, dissipation_rates, dissipation_format, axial_symmetry=True)
  File "/Users/snarayan/soft/QFI-Opt/spin_models.py", line 92, in simulate_sensing_protocol
    state_2 = evolve_state(state_1, time_2, entangling_hamiltonian, dissipator)
  File "/Users/snarayan/soft/QFI-Opt/spin_models.py", line 219, in evolve_state
    final_vec = solution.y[:, -1]
TypeError: list indices must be integers or slices, not tuple
```

Setting `USE_JAX = True` results in NaNs everywhere.

**Comment by perlinm** [comment]
@sriharikrishna that can be fixed by adding 
```python
    if time == 0:
        return density_op
```
On line 187 of `spin_ops.py` (that is, as the first thing in `evolve_state`).

**Comment by perlinm** [comment]
Also, looks like the JAX jacobian is wrong on that commit as well:
```bash
(QFI-Opt) [11:48:14] ~/super.tech/QFI-Opt (432d440)
$ python spin_models.py --num_qubits 1 --params 0.5 0.0 0.0 0.0

d(final_state/d(params[0]):
[[ 1.57079633+0.j          0.        -1.57079633j]
 [ 0.        -1.57079633j -1.57079633+0.j        ]]
```

**Comment by sriharikrishna** [comment]
I am able to recreate the bug, in the `compare-jacobians` branch using the JAX primal for finite differences

**Comment by perlinm** [comment]
I also tried this on
```bash
commit 8e48d1e509cde07f9519d73e63864ba7786a346a (HEAD)
Author: Michael A. Perlin <mika.perlin@gmail.com>
Date:   Tue Apr 11 11:35:53 2023 -0500

    get autodiff to work
```
which is 2 commits and ~1 hour after the first commit (`9d2b44d86bc1c147193b7398df8d474d25a01b41`) that added `ode_jax.py`.

The jacobian is wrong here too (after adding a similar `if time == 0: return density_op` fix):
```bash
(QFI-Opt) [11:57:00] ~/super.tech/QFI-Opt (8e48d1e) [!]
$ python spin_models.py --num_qubits 1 --params 0.5 0.0 0.0 0.0

d(finalstate[ 0 , 0 ])/d(params)=  [1.57079633+0.j 0.        +0.j 0.        +0.j 0.        +0.j]
d(finalstate[ 0 , 1 ])/d(params)=  [0.-1.57079633j 0.+0.j         0.+0.j         0.+0.j        ]
d(finalstate[ 1 , 0 ])/d(params)=  [0.-1.57079633j 0.+0.j         0.+0.j         0.+0.j        ]
d(finalstate[ 1 , 1 ])/d(params)=  [-1.57079633+0.j  0.        +0.j  0.        +0.j  0.        +0.j]
[<X>, <Y>, <Z>]: [Array(0., dtype=float64), Array(1., dtype=float64), Array(-2.22044605e-16, dtype=float64)]
[var(X), var(Y), var(Z)]: [Array(1., dtype=float64), Array(0., dtype=float64), Array(1., dtype=float64)]
```
In particular, `d(finalstate[ 0 , 1 ])/d(params)` and `d(finalstate[ 1 , 0 ])/d(params)` should be zero vectors.  So it looks to me like `ode_jax.py` was always wrong.

**Comment by sriharikrishna** [comment]
Addressed by PR #39
